### ISSUE TITLE
Bug: Use cli-driven workflow to run plan in VCS workspace #644

### ISSUE THREAD

**Issue by ilkerispir** [AzBuilder/terrakube#661]
### Bug description 🐞

Hello @alfespa17, @jcanizalez;

The remote plan is triggered from the local machine but fails because it does not download the modules.

Example Screenshots:

Fail plan from my localhost:
![image](https://github.com/AzBuilder/terrakube/assets/37123041/27f7f00d-228b-494b-a940-3c9ff4b3bf8c)

Success plan from Terrakube UI:
![image](https://github.com/AzBuilder/terrakube/assets/37123041/c3d4e6a0-d2b2-4353-a4a4-9f9cfb21cadf)

Related PR: https://github.com/AzBuilder/terrakube/pull/644

### Steps to reproduce

-

### Expected behavior

-

### Example repository

-

### Anything else?

-

**Comment by alfespa17** [comment]
Not sure how to reproduce this one.

I just add the backend.tf with the following locally:

```
terraform {
  cloud {
    organization = "simple"
    hostname = "8080-azbuilder-terrakube-srvv8ms68ej.ws-us107.gitpod.io"

    workspaces {
      name = "simple-terraform"
    }
  }
}
```

After that I can run locally the following:

```
terraform init
terraform plan
```

And it is working as expected, it should trigger a new job using the local changes....

How did you tested it?

**Comment by ilkerispir** [comment]
@alfespa17, it's almost the same, but as I said, my workspace has a dependency on AWS, for example. There is no problem in a scenario where there is no dependency such as AWS. I wonder if you have no problems if you try it the way I shared below?

```
terraform {
  required_version = ">= 1.4.6"

  required_providers {
    aws = {
      source                = "hashicorp/aws"
      version               = ">= 3.41.0"
      configuration_aliases = [aws.us_east_1]
    }
  }

  cloud {
    hostname     = "*"
    organization = "*"

    workspaces {
      name = "*"
    }
  }
}
```

**Comment by alfespa17** [comment]
I tried with the following:

```
terraform {

  required_providers {
    aws = {
      source                = "hashicorp/aws"
      version               = ">= 3.41.0"
    }
  }


  cloud {
    organization = "simple"
    hostname = "8080-azbuilder-terrakube-qx71l6xfqz6.ws-us107.gitpod.io"

    workspaces {
      name = "simple-terraform"
    }
  }
}
```
And it work for me running a plan after running ***"terraform init -upgrade"***

```
user@pop-os:~/git/simple-terraform$ terraform init -upgrade

Initializing Terraform Cloud...
Upgrading modules...
- time_module in module

Initializing provider plugins...
- Finding hashicorp/aws versions matching ">= 3.41.0"...
- Finding latest version of hashicorp/null...
- Finding latest version of hashicorp/time...
- Finding latest version of hashicorp/random...
- Using previously-installed hashicorp/null v3.2.2
- Using previously-installed hashicorp/time v0.10.0
- Using previously-installed hashicorp/random v3.6.0
- Installing hashicorp/aws v5.31.0...
- Installed hashicorp/aws v5.31.0 (signed by HashiCorp)

Terraform has made some changes to the provider dependency selections recorded
in the .terraform.lock.hcl file. Review those changes and commit them to your
version control system if they represent changes you intended to make.

Terraform Cloud has been successfully initialized!

You may now begin working with Terraform Cloud. Try running "terraform plan" to
see any changes that are required for your infrastructure.

If you ever set or change modules or Terraform Settings, run "terraform init"
again to reinitialize your working directory.
user@pop-os:~/git/simple-terraform$ terraform plan

Running plan in Terraform Cloud. Output will stream here. Pressing Ctrl-C
will stop streaming the logs, but will not stop the plan running remotely.

Preparing the remote plan...

To view this run in a browser, visit:
https://8080-azbuilder-terrakube-qx71l6xfqz6.ws-us107.gitpod.io/app/simple/simple-terraform/runs/5

Waiting for the plan to start...

***************************************
Running Terraform PLAN
***************************************
null_resource.previous: Refreshing state... [id=6529404432382828862]
module.time_module.random_integer.time: Refreshing state... [id=3]
time_sleep.wait_30_seconds: Refreshing state... [id=2024-01-03T21:25:13Z]
null_resource.next: Refreshing state... [id=3760039172153540685]

Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # null_resource.next2 will be created
  + resource "null_resource" "next2" {
      + id = (known after apply)
    }

  # null_resource.next3 will be created
  + resource "null_resource" "next3" {
      + id = (known after apply)
    }

Plan: 2 to add, 0 to change, 0 to destroy.

Changes to Outputs:
  + fake_data2    = "inputFromUI"

```

**Comment by adamlc** [comment]
@ilkerispir did you manage to solve this problem? I am having when importing an existing VCS project and state.

I don't have any problems however when I'm using a new VCS project. I've tried deleting lock files etc

**Comment by adamlc** [comment]
So after some research and experimentation I've determined the cause. I have a workaround, but I'm not sure if it's actually a bug or not.

Basically I have the following folder structure. I have 4 environments in a single repo and 1 shared modules folder in the root.

```
├── envs
│   ├── development
│   │   ├── main.tf
│   ├── mirror
│   │   ├── main.tf
│   ├── non-production
│   │   ├── main.tf
│   └── production
│       ├── main.tf
├── modules
│   └── base
│       ├── main.tf
```

So based on the above my working directory for development in terrakube is `/envs/development` for example.

When I do plans in the terrakube UI they work perfectly referencing the modules directory outside of the working directory. But when I run a plan from the CLI I get the errors about required plugins as above.

If I move the modules directory into the development folder or create a symlink, then the plan from the CLI works. So it sounds like it could be an issue with referencing a folder or something outside of the working directory?

I am not sure if the problem is the same one as above. But it might be a good starting point for a PoC bug

**Comment by alfespa17** [comment]
Interesting 🤔, based on the above when you are using the CLI driven workflow to trigger the plan for a VCS workspace the Terraform CLI is creating a ZIP from the folder where you are running the plan and upload the zip file to terrakube, internally terrakube use the content of the zip file to execute the plan, I guess the zip file migrh not have your modules from the root folder.

Could you try to go to the workspace setting and change the execution mode from remote to local, run the Terraform init and try to run the plan again.

In this way the plan will run locally using your current folder structure.

**Comment by ilkerispir** [comment]
Good catch @adamlc, thank you. 

@alfespa17, yes, it is exactly as @adamlc said. When I downloaded the terraformContent.tar.gz file of the relevant job from S3, unfortunately, there were no modules on which the environment depended. There is only the root directory in which it is running. That's why we are getting the error. When I deleted the dependencies the plan worked successfully. But this time, since I deleted the dependencies, it wants to destroy the dependencies I deleted. If the local module dependencies are also uploaded in the uploaded tar.gz file, the problem will be solved. As I said before, this feature is available in Terraform Cloud with a remote working plan.

**Comment by adamlc** [comment]
Great well glad we've gotten to the bottom of it. It also works on [OTF](https://github.com/leg100/otf). Not sure if there is anything useful in the old source code there that helps.

**Comment by alfespa17** [comment]
I will try to replicate using the above folder structure so I can replicate the issue and check if I can find the solution, 

I will update the issue in the following days

**Comment by ilkerispir** [comment]
Hello @alfespa17, is there any progress on this issue? also is there anything I can help you with? I could push a fix commit for this issue. However, my Java knowledge is limited.

**Comment by alfespa17** [comment]
I think I was able to fix this issue, now when I run the CLI I can see the message in PR #756 

```
The remote workspace is configured to work with configuration at
/work1 relative to the target repository.

Terraform will upload the contents of the following directory,
excluding files or directories as defined by a .terraformignore file
at /home/user/git/webhook/simple-terraform/.terraformignore (if it is present),
```

The terraformContent.tar.gz now include all the information.

![image](https://github.com/AzBuilder/terrakube/assets/4461895/d8a003e0-d485-4296-b84b-85e6fdec83f5)


Maybe when you have some free time you could test it @ilkerispir @adamlc and thank you for helping me find the root cause of the issue

You could use version 2.20.0-beta.7

**Comment by ilkerispir** [comment]
Thank you very much, @alfespa17. I will try it as soon as possible.

**Comment by adamlc** [comment]
@alfespa17 thanks for the fix. Whilst the original error has gone away, I'm now getting errors about undeclared variables.

I have about 10 variables defined in terrakube. But for some reason it looks like variables in my variables.tf aren't being picked up.

I haven't had much time to do any debugging yet, but just thought I'd point it out!

```
���
��� Error: Value for undeclared variable
���
��� A variable named "twilio_auth_token" was assigned on the command line, but
��� the root module does not declare a variable of that name. To use this
��� value, add a "variable" block to the configuration.
���
���
��� Error: Value for undeclared variable
���
��� A variable named "twilio_sid" was assigned on the command line, but the
��� root module does not declare a variable of that name. To use this value,
��� add a "variable" block to the configuration.
���
```

**Comment by alfespa17** [comment]
I think I know the reason for this one 🤔, I will check that later today

**Comment by alfespa17** [comment]
I suspect the problem was in the working directory where terraform was executing, I did an small fix in [2.20.0-beta8](https://github.com/AzBuilder/terrakube/releases/tag/2.20.0-beta.8)
